src/operator/math_functions-inl.h

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file special_functions-inl.h
+ * \brief
+ * \author Matthias Seeger
+*/
+
+
+#ifndef MXNET_OPERATOR_MATH_FUNCTIONS_INL_H_
+#define MXNET_OPERATOR_MATH_FUNCTIONS_INL_H_
+
+#include <cmath>
+
+namespace mxnet {
+namespace op {
+
+namespace math {
+
+#define MATH_FUNC1(name) \
+template<typename DType> inline \
+DType name(DType a) { \
+  return DType(std::name(static_cast<float>(a))); \
+} \
+template<> inline \
+float name(float a) { \
+  return std::name(a); \
+} \
+template<> inline \
+double name(double a) { \
+  return std::name(a); \
+} \
+template<typename DType> inline \
+float name##f(DType a) { \
```

Comment:
Why do we want this? Looks weird

Reply:
The idea is that whenever any math.h function is used in an expression with DType, I compute it internally in float, and cast back to DType in the end. This is why the ***f variants are needed here: inside an expression, you do not want to cast back to DType, only at the end.